793 Safe MPI collectives

lifflander opened the pull request:
Fixes #793 

Implement a new call to enqueue an MPI collective and run it safely **while** VT is running.

```C++
auto tag = vt::theCollective()->mpiCollective([]{
   MPI_Barrier(MPI_COMM_WORLD);
});

vt::theCollective()->waitCollective(tag);
```

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -0,0 +1,140 @@
+  int done = 0;
+
+  // These three collective can execute in any order, but it will always be
+  // consistent across all the nodes
```
I think this comment was incorrectly copied from above

lifflander replied:
Fixed

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -0,0 +1,140 @@
+  // consistent across all the nodes
+  auto tag = theCollective()->mpiCollective([&done]{
+    auto comm = theContext()->getComm();
+    int val = 0;
```
For a stronger test, initialize `val` to the present rank, and then assert after the `waitCollective` that they all end up with `val == 0`

lifflander replied:
I've strengthened the test based on this idea

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -0,0 +1,140 @@
+  theCollective()->mpiCollectiveWait([&done]{
+    auto comm = theContext()->getComm();
+    int val_in = 1;
+    int val_out = 0;
```
Similarly, it would be good to have an assertion that this comes out to equal `Comm_size`

lifflander replied:
Done

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,69 @@ struct CollectiveAlg :
 
+public:
+  /**
+   * \brief Enqueue a lambda with an embedded MPI collective invocation to
```
Is it strictly
* One MPI collective?
* Just collectives? 

I would think the actual protocol should be equally fine with any closed set of MPI operations. I.e. as long as the function isn't leaving any incomplete requests around, anything should go.

For the Zoltan use case, this is kind of an important distinction, since the call to Zoltan is not in itself 'an MPI collective' - it makes some number of MPI calls, collective, p2p, or otherwise.

lifflander replied:
I agree it's any closed set that is allowed that doesn't bleed out.

lifflander replied:
I've improved the documentation to state a closed set of MPI collectives

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,69 @@ struct CollectiveAlg :
+  /**
+   * \brief Enqueue a lambda with an embedded MPI collective invocation to
+   * execute in the future. Returns immediately, enqueuing the action for the
+   * future.
```
There's also a potential need to document the buffer semantics across this sort of thing. Essentially, any buffer referenced inside `action` needs to be treated as part of an active MPI request until the `tag` is waited upon or `isCollectiveDone` returns true  for it.

lifflander replied:
I've added some documentation for this

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,69 @@ struct CollectiveAlg :
+   *
+   * \return tag representing the collective
+   */
+  TagType mpiCollective(ActionType action);
```
I think it would be preferable to name this `mpiCollectiveAsync` to be explicit about it.

lifflander replied:
Ok, I've renamed this `mpiCollectiveAsync`

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,69 @@ struct CollectiveAlg :
+   *
+   * \return tag representing the collective
+   */
+  TagType mpiCollective(ActionType action);
```
It would be nice if we could declare this `[[nodiscard]]`, but that's C++17 and the compiler support isn't there yet. If you have a macro for it in VT, it should be applied here.

lifflander replied:
You are allowed to discard by calling this in an epoch and waiting for termination of that epoch.

I was thinking that maybe I missed a produce/consume in postponed collectives, but since I recurse on the stack it shouldn't be able to complete the epoch until the message buffer is empty.

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,69 @@ struct CollectiveAlg :
+   *
+   * \return tag representing the collective
+   */
+  TagType mpiCollective(ActionType action);
```
How are callers running from asynchronous handlers supposed to ensure that they're calling this in the same sequence everywhere? I assume that's necessary to ensure that each action is associated with the same tag on all processes

PhilMiller replied:
Regardless, if that is a requirement, it needs to be documented

lifflander replied:
Yes, that is a requirement. It must be called in the same order everywhere. I wanted to relax this requirement, but couldn't figure out a cheap/safe and relatively easy way.

I started with a `std::string` parameter to match them up. But, I couldn't figure out how to make that API performant and correct. I hashed the `std::string`, which is fine, but collisions quickly become an issue. I can detect a collisions easily enough on a given node, but resolving that collision is a problem. If I allow any arbitrary order with `std::string` tags, I can't ever hold anything back because that could deadlock due to a bad ordering with a colliding string

PhilMiller replied:
You could require the callers to provide a unique integer tag, instead of assigning it internally. They're much better positioned to be able to generate them properly. Either there's actually one stream of events that generate these calls, or there are multiple sources, and they need to partition the space.

There's an intermediate, of offering a 'sequence' facility, in which distinct call sources each pass a different value indicating what 'sequence' the call is part of, and then you just have an incrementing counter for each.

lifflander replied:
What do you think about this:
```C++
auto scope = theCollective()->createCollectiveScope(/*integer tag*/ tag);
scope.mpiCollectiveAsync([]{}); //ordered within scope
scope.mpiCollectiveAsync([]{}); //ordered within scope
```

Where `scope` is a move-only object that the user can hold on (e.g., inside an objgroup). Within scope all operations are collectively ordered. Each scope has its own sequence.

PhilMiller replied:
Yes, that's pretty much what I had in mind. Much easier to uniquely identify the scopes, and then sequence on each one.

Are you going to just make the keys `pair<scope_tag, sequence_in_scope>` then?

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+}
+
+void CollectiveAlg::waitCollective(TagType tag) {
+  while (not isCollectiveDone(tag)) {
```
Does this very soon need to be the nested scheduler call? Just flagging an upcoming semantic merge conflict

lifflander replied:
Yes, good point

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+  // stack once!
+  static int reenter_counter_ = 0;
+
+  if (reenter_counter_ == 0) {
```
As with many other instances, could you put the `else` case up front, ending it with a return, and un-nest the rest of the function?

lifflander replied:
Done

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+
+  NodeType collective_root = 0;
+
+  auto cb = theCB()->makeBcast<CollectiveMsg,&runCollective>();
```
I forget - do vt broadcasts guarantee ordered delivery? If so, it would be helpful to put a comment to that effect here, since it seems like this code relies on that being the case.

PhilMiller replied:
Ok, so it doesn't seem like the code actually relies on ordering, since it goes with whatever the root is running. This just needs prose ahead of the code describing the reasoning involved.

lifflander replied:
Ok, I've added a comment to explain this.

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+    collective_root, msg.get(), cb, tag, no_seq_id, 1, ident
+  );
+
+  next_tag_++;
```
Why not put this increment up next to the read from `next_tag_`? When I saw that above, I wondered "where's the increment? Is it missing?"

lifflander replied:
Ok, I've moved it up

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+
+    reenter_counter_--;
+
+    // Recurse, running a postponed collective
```
If you have ordered vt broadcast, LIFO ordering on `postponed_collectives_` can get you mis-matched collectives, since you're not guaranteed that every process has received the same last broadcast.

Even if this gets switched to FIFO, ordered delivery doesn't ensure that every process has received the broadcast carrying the 'next' collective yet. I think this needs to do an `MPI_Allreduce(&postponed.front->tag, &min_tag, MPI_MIN)`, everyone checks `bool matching = min_tag == my_front_tag;`, and then `MPI_Allreduce(matching, MPI_LAND)` to make sure every process agrees

PhilMiller replied:
OK, I see the logic a little better now - we can only reach this point because of a broadcast, that's the callback from a reduction, in which every process contributed (ostensibly) the same tag. So, we're guaranteed up front that every process has that tag and the corresponding action.

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+
+    // We need a barrier here so the root doesn't finish the broadcast first and
+    // then enter the collective. We could replace the Ibcast/Ibarrier with a
+    // Iallreduce, but I think this is cheaper
```
It's not really going to be cheaper, but it is easier to reason out the logic that everyone has the same value for `consensus_tag`, if it's coming from a broadcast

lifflander replied:
Yes, I agree. The Iallreduce logic (which I started with) is definitely tricker.

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,141 @@ CollectiveAlg::CollectiveAlg()
+      auto iter = planned.find(msg->tag_);
+      vtAssert(iter != planned.end(), "Planned collective does not exist");
+      consensus_tag = msg->tag_;
+      action = iter->second.action_;
```
You can omit checking `planned` and pulling out `action` here, and then just make the code below do it unconditionally, rather than only on non-root. Everything up to that point is only concerned with determining `consensus_tag`

lifflander replied:
Good point

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,77 @@ struct CollectiveAlg :
 
+public:
+  /**
+   * \brief Enqueue a lambda with an embedded closed set of MPI collectives
```
The operation isn't limited to MPI collectives, either. Just MPI operations.

lifflander replied:
Changed to operations

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,77 @@ struct CollectiveAlg :
+   *
+   * Any buffers captured in the lambda to use with MPI collective operations
+   * need to exist until \c isCollectiveDone returns \c true or \c
+   * waitCollective returns on the returned \c TagType
```
It's not just that the buffers need to exist, it's that as far as MPI operation is concerned, they need to be considered 'in use' (I don't remember if that's the phrasing from the standard) - other code can't read or write to them, without screwing up results.

lifflander replied:
Ok, fixed

PhilMiller commented on src/vt/collective/collective_alg.h:
```diff
@@ -83,6 +83,77 @@ struct CollectiveAlg :
+  /**
+   * \brief Enqueue a lambda with an embedded closed set of MPI
+   * collectives. Spin in the VT scheduler until it terminates. All other nodes
+   * must not depend on this node to enqueue a matching collective.
```
Given the scheduler spinning, is the rule against cross-node dependencies actually necessary?

lifflander replied:
I think you're right, it isn't necessary. As long as the user has messages in flight that might enable the collective elsewhere, it should be fine.

PhilMiller commented on src/vt/collective/collective_alg.cc:
```diff
@@ -52,4 +52,148 @@ CollectiveAlg::CollectiveAlg()
+    consensus_tag, msg->tag_
+  );
+
+  // The root had a different collective in mind... let's do that one
```
Repeating my earlier comment - The code can just declare and look up `action` here, unconditionally. No need for the root to pick it out earlier, or to conditionalize this.

PhilMiller replied:
By making that change, you also then don't need to repeat `planned.find(consensus_tag)` below in the `erase` call

lifflander replied:
Ok, done

PhilMiller commented:
OK, protocol-wise, I'm satisfied that this works as intended. I think there may actually be a short workshop paper in this, since the design should be useful for Charm++, HPX, etc. "Embedding distributed MPI operations in asynchronous tasking runtime applications" or something

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

PhilMiller commented:
That all looks great.

I'm 90% convinced that the protocol as implemented guarantees that every call made does eventually get completed. The vt reductions and broadcasts are guaranteed to make progress, resulting in every process seeing as many handler calls as there were operation requests. For any given request, on any given process, either the process will be at `reenter == 0`, and go into the consensus protocol, or it will be non-zero, and will re-start the consensus protocol for it when the current operation is complete. So, every process eventually executes as many rounds of the consensus protocol, and hence that many actions, as calls were made.

Also, this will end up having interesting interactions with a hypothetical PMPI tracing into vt's log files implementation.

lifflander commented:
Travis CI seems to be screwed up---those build failures are spurious.

lifflander commented:
@PhilMiller I think I've found a theoretical bug. If two mpiCollectiveAsync operations are enqueued in different epochs, and inside those operations messages are sent from VT, they might get the wrong epoch because I don't push/pop the epoch when recursing on the stack.

PhilMiller commented:
> @PhilMiller I think I've found a theoretical bug. If two mpiCollectiveAsync operations are enqueued in different epochs, and inside those operations messages are sent from VT, they might get the wrong epoch because I don't push/pop the epoch when recursing on the stack.

I think you may be better off disallowing vt message sends during these operations, or set things up in a way that they have to push/pop their own epoch explicitly, with a failure if they don't

lifflander commented:
Should I implement the scoping as discussed in that comment in this PR? Shouldn't be too difficult to do.

PhilMiller commented:
May as well get the scoping in now, sure

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
@PhilMiller I've adding in scopes. Ended up being more complex than I thought to get the desired behavior. Not sure it's worth the complexity, let me know what you think.

If we keep this I need to write more tests.

You can call `auto scope = theCollective()->makeCollectiveScope()` and each time get a new distinct system-tagged scope.

You can call it with an explicit tag `auto scope = theCollective()->makeCollectiveScope(5)`, and get user-tagged scope that won't conflict with generated system scopes.

When a `scope` goes out of scope (move-only), you can't add any more operations. It doesn't wait on everything (considered that design but I think its too tricky), but it does cleanup when it finishes.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

PhilMiller commented on src/vt/collective/collective_scope.h:
```diff
@@ -0,0 +1,174 @@
+
+private:
+  bool is_user_tag_ = false;
+  TagType scope_ = no_tag;
```
Why duplicate the tag between the public interface and the underlying implementation?

lifflander replied:
Good point. I don't need it in `ScopeImpl`, just here to look up the scope from the `CollectiveScope`

lifflander replied:
Fixed

PhilMiller commented on tests/unit/collectives/test_mpi_collective.cc:
```diff
@@ -151,5 +151,57 @@ TEST_F(TestMPICollective, test_mpi_collective_3) {
+  }
+}
+
+
```
Could this test be adapted to actually affirm that the MPI operations were run in the same order everywhere?

lifflander replied:
Yes, I just added that!

PhilMiller commented:
I think the scopes look good. In the tests, it may be worthwhile to spell out `vt::collective::CollectiveScope`, since actual users will need to declare member variables of that type.

lifflander commented:
> vt::collective::CollectiveScope

Do you think I should alias `CollectiveScope` in the `vt` namespace directly?

lifflander commented:
I have spelled it out in the tests now.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
@PhilMiller

- I left `CollectiveScope` in `collective::` for now without an alias.
- I spelled it out in the tests
- I added a consistent order check across nodes for the last test
- Removed duplicate/unneeded scoping info in `ScopeImpl`

I think this is ready to go!

The only last thing would be to start using the interface across the system, but I think that should go in a seperate PR. Maybe starting with `group/collective/group_info_collective.cc`.

lifflander commented:
Looks like there might be non-deterministic failure on Travis that is real.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
Removing that barrier caused a bunch of other failures:

```
The following tests FAILED:
1673	  7 - vt:*/TestActiveBroadcastPut.test_type_safe_active_fn_bcast2/*_proc_2 (Timeout)
1674	  8 - vt:*/TestActiveBroadcast.test_type_safe_active_fn_bcast2/*_proc_2 (Timeout)
1675	 66 - vt:*/TestSequencerExtensive.test_wait_1/*_proc_2 (Timeout)
1676	 73 - vt:*/TestSequencerParallelParam.test_seq_parallel_param/*_proc_2 (Timeout)
1677	 91 - vt:*/TestTermGlobal.test_term_detect_routed/*_proc_2 (Timeout)
```

lifflander commented:
I think I have the problem solved now. We need the collective barrier, but not the MPI one after.

PhilMiller commented:
Could you explain why, and what went wrong?

On Fri, May 1, 2020, 18:26 Jonathan Lifflander <notifications@github.com>
wrote:

> I think I have the problem solved now. We need the collective barrier, but
> not the MPI one after.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DARMA-tasking/vt/pull/795#issuecomment-622592394>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA64ACMYGB7MV57GURFOSDRPNEADANCNFSM4MTL4YJQ>
> .
>

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collectives/test_mpi_collective.cc  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5442758&bid=17847806)

lifflander commented:
> Could you explain why, and what went wrong?
> […](#)

I changed this line to:

```C++
theCollective->barrierThen([]{
  MPI_Barrier(comm);
});
```

Just a regular, blocking VT barrier.

```C++
theCollective->barrier();
```

The regular barrier is needed and safe. When VT re-inits the runtime, which is used in some tests, it calls setup to synchronize.

lifflander commented:
@PhilMiller
